libs/cri-api-service/src/main/java/uk/gov/di/ipv/core/library/criapiservice/CriApiService.java

```diff
@@ -358,4 +384,37 @@ private VerifiableCredentialResponse getVerifiableCredentialResponseForApplicati
         }
         return vcResponseBuilder.build();
     }
+
+    @Tracing
+    public Optional<JWK> getKeyFromJwksEndpoint(URI jwksEndpoint) {
+        try {
+            LOGGER.info(LogHelper.buildLogMessage("Retrieving key from well-known endpoint."));
+            var request = HttpRequest.newBuilder().uri(jwksEndpoint).GET().build();
+            var httpResponse = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+
+            if (httpResponse.statusCode() != SC_OK) {
+                LOGGER.warn(
+                        LogHelper.buildLogMessage(
+                                String.format(
+                                        "%d: Error retrieving public encryption key.",
+                                        httpResponse.statusCode())));
+                return Optional.empty();
+            }
+
+            var keys = JWKSet.parse(httpResponse.body()).getKeys();
+
+            return keys.stream().filter(key -> key.getKeyUse() == KeyUse.ENCRYPTION).findFirst();
```

Comment:
At the moment the method says 'getKeyFromJwksEndpoint' but this is specifically returning the first encryption key (not an arbitrary key).

I'd suggest making this method return the entire JWKS object, and then consumers can decide which key they want. That's probably also what we want to store in our cache too.